### ISSUE TITLE
Increment timestep raising error: Unknown attribute 'where' of type 'module 'xarray.'

### ISSUE THREAD

**Issue by imscw95** [EcohydrologyTeam/ClearWater-modules#74]
![xr where module error](https://github.com/EcohydrologyTeam/ClearWater-modules/assets/66634423/c8e6142b-8fd6-49c5-a93f-acb921597816)

**Comment by sjordan29** [comment]
Looking at #39, it looks like functions with `xr.where()` should be used without `@numba.njit`.  Removing that moves to the next error.

Additional related issues to address as you work through:
* `PAR`: need the failure condition in `xr.where` 
* Functions with multiple conditions (e.g., `FL`, `FN`, `FP`, `mu` -- likely more): Use `np.select()` as described in Issue #39 and PR #44. I believe you can keep `numba` when you use `np.select`
* For `mu`: something funky is going on where `mu` is not a numeric object

Let me know if you need help with making any of these changes!

**Comment by imscw95** [comment]
Ok thank you. I will look at all of those!

**Comment by imscw95** [comment]
So I am unable to see all the comments in https://github.com/EcohydrologyTeam/ClearWater-modules/pull/44 for some reason, though I remember a discussion about how to use `np.select` in the processes. I will try to use `np.select` without referencing those comments, though my one concern is making sure I am applying it correctly over the full `xr.DataArray` grid.

**Comment by sjordan29** [comment]
Sounds good -- looks like there are some examples in the [TSM processes](https://github.com/EcohydrologyTeam/ClearWater-modules/blob/main/src/clearwater_modules/tsm/processes.py) too if those are helpful to see

**Comment by imscw95** [comment]
Oh yes fantastic thanks. I see that `np.where` is used in TSM, and `xr.where` is used in NSM. Not sure if that's a problem but I assume we'd want to use one or the other and not both. Thoughts?

**Comment by sjordan29** [comment]
Good call! yeah it looks like `np.select()` and `np.where()` are both used in TSM in place of `xr.where()`. I know some testing revealed `np.select()` is faster than `xr.where()` for nested logic --  however, I don't see as much discussion on `np.where()` vs. `xr.where()` for simpler if/else logic. It does look from TSM like `np.where()` can be used with `numba` but not sure of other merits. My guess is the whole discussion in #57 may be relevant to this as well. 
@aufdenkampe were you part of that decision process / do you have any thoughts on `np.where()` vs `xr.where()`? we can do some more testing but if thought has already gone into this I don't want to duplicate any efforts.

**Comment by aufdenkampe** [comment]
@sjordan29 & @imscw95, I think that using `xr.where()` would be preferable to using `np.where()` in most cases. It is xarray native and likely to be [zero-copy](https://en.wikipedia.org/wiki/Zero-copy), which would be better for memory. It was also quite fast in Xavier's testing within [`running_process_funcs.ipynb`](https://github.com/EcohydrologyTeam/ClearWater-modules-python/blob/main/examples/dev_sandbox/running_process_funcs.ipynb)

So the only reason to use `np.where()` is when you have other reasons to apply `@numba.njit` to the function.

**Comment by sjordan29** [comment]
I think every process in TSM / NSM is currently using `@numba.njit`, which is probably why TSM used `np.where()`. However, examining the discussion in #57 suggests that we should start moving away from numba all together, perhaps starting with NSM which would allow using `xr.where()` as you suggest @aufdenkampe

**Comment by aufdenkampe** [comment]
@sjordan29. I'm a fan of moving away from `numba`.

@xaviernogueira's test results shown in https://github.com/EcohydrologyTeam/ClearWater-modules/issues/57#issuecomment-1856667822 make showed that the native `xarray` functions always performed faster than `numba`!!!  I provided additional thoughts in my followup https://github.com/EcohydrologyTeam/ClearWater-modules/issues/57#issuecomment-1964857463

**Comment by sjordan29** [comment]
Great! @imscw95, then I'd recommend proceeding with `xr.where()` for simple if/else logic and `np.select()` for more complicated logic, and removing `numba` from NSM processes. We can open a separate issue to address this in in TSM.

**Comment by sjordan29** [comment]
I added issue #76 for TSM.

**Comment by imscw95** [comment]
Replaced all nested `xr.where` with `np.select` and committed the changes in the IJM_NSM_nb branch. When I run increment_timestep, I am seeing an error when comparing an np.ndarray type to a float type - this pops up for the first process that features 2 `np.select` functions (FL). However, this appears to be the same set up as the richardson's number computation for TSM in terms of comparing types, and using two np.select functions in one process. I will come back to this but wanted to raise the issue for everyone's awareness.

**Comment by imscw95** [comment]
Feel free to work on it, I have to work on some other stuff for the rest of the day.

**Comment by sjordan29** [comment]
I committed some updates. The specific issue you were running into here was that in `np.select()`, the default argument was set to the string `"NaN"` when we should use `np.nan`. Otherwise, there's some confusion over mixed float/string arrays. 

Some additional updates I made while troubleshooting:
* `np.exp()` plays better with arrays than `math.exp`. 
* some `condlist` and `choicelist` inputs were swapped for `np.select()` which throws an error

**Comment by imscw95** [comment]
Are we removing numba from all process functions or just where we are using xr.where?

**Comment by sjordan29** [comment]
Great! yeah remove it from all process functions for NSM.
